runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/testing/TestDataflowPipelineRunnerTest.java

```diff
@@ -378,4 +383,251 @@ public State answer(InvocationOnMock invocation) {
+    PAssert.that(pc).containsInAnyOrder(1, 2, 3);
+
+    final DataflowPipelineJob mockJob = Mockito.mock(DataflowPipelineJob.class);
+    when(mockJob.getDataflowClient()).thenReturn(service);
+    when(mockJob.getState()).thenReturn(State.DONE);
+    when(mockJob.getProjectId()).thenReturn("test-project");
+    when(mockJob.getJobId()).thenReturn("test-job");
+
+    DataflowPipelineRunner mockRunner = Mockito.mock(DataflowPipelineRunner.class);
+    when(mockRunner.run(any(Pipeline.class))).thenReturn(mockJob);
+
+    TestDataflowPipelineRunner runner = (TestDataflowPipelineRunner) p.getRunner();
+    p.getOptions().as(TestPipelineOptions.class)
+        .setOnCreateMatcher(new BaseMatcher<PipelineResult>() {
+          @Override
+          public boolean matches(Object o) {
+            // Assert that job.WaitToFinish has not yet been called
+            try {
+              verify(mockJob, never()).waitToFinish(any(Long.class), any(TimeUnit.class),
+                  any(JobMessagesHandler.class));
+            } catch (Exception e) {
+              return false;
+            }
+            return (PipelineResult) o == (PipelineResult) mockJob;
+          }
+
+          @Override
+          public void describeTo(Description description) {
+          }
+        });
+    when(mockJob.waitToFinish(any(Long.class), any(TimeUnit.class), any(JobMessagesHandler.class)))
+        .thenReturn(State.DONE);
+
+    when(request.execute()).thenReturn(
+        generateMockMetricResponse(true /* success */, true /* tentative */));
+    runner.run(p, mockRunner);
+  }
+
+  @Test
+  public void testBatchSuccessMatcherPipelineSuccess() throws Exception {
+    Pipeline p = TestPipeline.create(options);
+    PCollection<Integer> pc = p.apply(Create.of(1, 2, 3));
+    PAssert.that(pc).containsInAnyOrder(1, 2, 3);
+
+    final DataflowPipelineJob mockJob = Mockito.mock(DataflowPipelineJob.class);
+    when(mockJob.getDataflowClient()).thenReturn(service);
+    when(mockJob.getState()).thenReturn(State.DONE);
+    when(mockJob.getProjectId()).thenReturn("test-project");
+    when(mockJob.getJobId()).thenReturn("test-job");
+
+    DataflowPipelineRunner mockRunner = Mockito.mock(DataflowPipelineRunner.class);
+    when(mockRunner.run(any(Pipeline.class))).thenReturn(mockJob);
+
+    TestDataflowPipelineRunner runner = (TestDataflowPipelineRunner) p.getRunner();
+    p.getOptions().as(TestPipelineOptions.class)
+        .setOnSuccessMatcher(new BaseMatcher<PipelineResult>() {
+          @Override
+          public boolean matches(Object o) {
+            // Assert that job.WaitToFinish has not yet been called
+            try {
+              verify(mockJob, Mockito.times(1)).waitToFinish(any(Long.class), any(TimeUnit.class),
+                  any(JobMessagesHandler.class));
+            } catch (Exception e) {
+              return false;
+            }
+            return (PipelineResult) o == (PipelineResult) mockJob;
```

Comment:
Set the test expectations different from the functional expectations of the code

assertSame(mockJob, o);
return true;